### PR TITLE
Decouple spell casting from speech controls

### PR DESCRIPTION
**Pull request by Foxinatel**
Addresses #44
This commit decouples the logic for spell casting from the game logic. This gives us extra granularity that allows us to much more easily reuse the speech plugin from other parts of the code, which can be used for things like menus.

### CODE REVIEW

**Review by AmosHolland — CHANGES_REQUESTED** [review]
This all looks good, as long as it's tested and working as intended I'm happy to make this change. Just have to wait for the updated controls to be integrated and then I'll be happy to merge.

**Comment by AmosHolland** [comment]
Looking at the new controls code this is going to need *significant* changes to integrate properly. In the new controls there's no defined point where recording stops, instead recording stops when the recogniser successfully recognises a word from it's grammar, just as a heads up.

**Comment by Foxinatel** [comment]
So the new system for managing speech and spell casting has been implemented, although this still poses some major conflicts. Instead of expensively storing our own buffer for audio data with large copies happening whenever we want to detect a word, we now stream the data into the vosk model in small snippets. This also allows us to make small partial queries to our model to see if it has detected a word since recording began.

Since this change from storing data to streaming it should implement a lot of the work done manually in #47, I think the ideal solution from here is to place a block on that PR until the remaining features become supported in this branch.

**Comment by Foxinatel** [comment]
We ball, the new spell casting system is fully implemented into the new speech system

File: src/spells.rs [hunk 1]
```diff
@@ -207,8 +206,20 @@ fn handle_straight_laser_traj_ind(
...
     gizmos.line(
-        t.translation(),
-        t.translation() + (t.forward() * ray_travel),
+        t.translation,
+        head_transform.translation + (head_transform.forward() * ray_travel),
         Color::RED,
     ); // TODO don't use gizmos for line drawing
 }
+
+fn track_spell_indictator(
```
Comment (rosefromthedead) [inline src/spells.rs:215]: optional: indictator -> indicator

**Review by rosefromthedead — APPROVED** [review]
LGTM, thanks!

File: src/spell_control.rs [hunk 2]
```diff
@@ -164,11 +142,86 @@ fn spawn_new_spell_entities(
...
+            .translation)
+        .length();
+
+    let mut spell_check_close = true;
+    for dist in [thumb_dists, index_dists] {
+        spell_check_close = spell_check_close && dist < 0.25;
+    }
+
+    spell_check_close
+}
```
Comment (AmosHolland) [inline src/spell_control.rs:227]: This function is defined twice, once here, and once in speech.rs, do we need to do this? Could we not just make one of them public?

**Review by AmosHolland — APPROVED** [review]

**Comment by AmosHolland** [comment]
Overall LGTM!

**Review by rosefromthedead — CHANGES_REQUESTED** [review]
Amos is so right, blocking this until `check_fingers_close` is dedup'd, otherwise someone's gonna have a hard time figuring out why changing the code in one of them has no effect on the callsite of the other

File: src/spell_control.rs (hunk 2 above)
Comment (Foxinatel) [reply src/spell_control.rs:227]: Yeah I was a little unsure about what to do with that. This decoupling means that the casting doesn't necessarily need to be the same action as stopping the recording. I think the shout is to put it in `speech.rs` as public.

File: src/spells.rs (hunk 1 above)
Comment (Foxinatel) [reply src/spells.rs]: fixed in most recent commit

**Review by AmosHolland — APPROVED** [review]
de-duping the function makes sense, lgtm!

**Comment by Foxinatel** [comment]
Had a bit of a nightmare getting these commits orderly, as this branch used commits from Alyssa that got redone, making this branch try to pull in a bunch of dead commits

File: src/spell_control.rs [hunk 3]
```diff
@@ -240,4 +192,4 @@ fn palm_mid_point_track(
         .translation;
 
     palms_mid_point_res.0 = left_palm.lerp(right_palm, 0.5);
-}
+}
```
Comment (rosefromthedead) [inline src/spell_control.rs:195]: Girl

**Review by AmosHolland — APPROVED** [review]
Looks good!

**Review by AmosHolland — APPROVED** [review]
Looks good!

**Review by rosefromthedead — APPROVED** [review]
I'm like semi conscious, but lgtm. Congrats on the narrowly negative diffstat :)

**Review by OwenBoisse — APPROVED** [review]
tested and lgtm

**Review by AmosHolland — APPROVED** [review]
still looks good!

**Pull request merged**